[inductor][fx] clarify padding logic

nmacchioni opened the pull request:
Summary:
`should_pad_common` and `should_pad_bench` logic were semi-intertwined which can make working with the padding logic difficult

previously there was no clear delineation as to what logic belonged in which of the two functions

to fix this, we introduce `can_pad` and `should_pad`.

`can_pad` contains all logic related to padding safety; is essence, can we pad this match without breaking anything

`should_pad` contains all logic related to padding performance (i.e. if we did apply padding, would it be performant)

this separation of can/should logic also makes caching simpler; we'd never want to cache "can" logic since it is 1) fast to compute and 2) messing up the "can" logic would only lead to issues

we also rename a previously named `should_pad` method to `is_padding_beneficial`, since that method strictly deals with checking if the padded option is sufficiently faster than the non-padded option

D88383059

@diff-train-skip-merge


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

v0i0 commented on torch/_inductor/fx_passes/pad_mm.py:
```diff
@@ -389,13 +456,13 @@ def should_pad_mm_bf16(dtype: torch.dtype, M: int, N: int, K: int) -> bool:
 
 
-def should_pad_bench(*args: Any, **kwargs: Any) -> bool:
+def should_pad(*args: Any, **kwargs: Any) -> bool:
```
i worry this might tempt someone later to only call should_pad without can_pad. i like the distinction, but how about we carry it through: i.e. these are still is_padding_beneficial functions, and should_pad functions call both can_pad and is_padding beneficial?

izaitsevfb commented:
@pytorchbot revert -c ghfirst -m "my bad, this is actually reverted internally"

nmacchioni committed:
[inductor][fx] clarify padding logic (#169577)

Summary:

`should_pad_common` and `should_pad_bench` logic were semi-intertwined which can make working with the padding logic difficult

previously there was no clear delineation as to what logic belonged in which of the two functions

to fix this, we introduce `can_pad` and `should_pad`.

`can_pad` contains all logic related to padding safety; is essence, can we pad this match without breaking anything

`should_pad` contains all logic related to padding performance (i.e. if we did apply padding, would it be performant)

this separation of can/should logic also makes caching simpler; we'd never want to cache "can" logic since it is 1) fast to compute and 2) messing up the "can" logic would only lead to issues

we also rename a previously named `should_pad` method to `is_padding_beneficial`, since that method strictly deals with checking if the padded option is sufficiently faster than the non-padded option

Test Plan:
```
# Primary test
buck test fbcode//mode/opt //caffe2/test/inductor:pad_mm

# Additional related tests
buck test fbcode//mode/opt //caffe2/test/inductor:padding //caffe2/test/inductor:deterministic //caffe2/test/inductor:autoheuristic
```

Reviewed By: v0i0

Differential Revision: D88383059

laithsakka commented on torch/_inductor/fx_passes/pad_mm.py:
am doing some changes on this file here to make better support unbacked 
and got questons

what is the else branch for? if a size is Not int and not SymInt what could it be lol?

laithsakka commented:
thanks for the refactor btw !
